lemmy_api/src/report.rs

```diff
@@ -0,0 +1,300 @@
+use actix_web::web::Data;
+use std::str::FromStr;
+
+use lemmy_db::{comment_report::*, comment_view::*, post_report::*, post_view::*, Reportable, ReportType,};
+use lemmy_structs::{blocking, report::*};
+use lemmy_utils::{APIError, ConnectionId, LemmyError};
+use lemmy_websocket::{LemmyContext, UserOperation, messages::SendUserRoomMessage};
+
+use crate::{check_community_ban, get_user_from_jwt, is_mod_or_admin, Perform};
+
+const MAX_REPORT_LEN: usize = 1000;
+
+#[async_trait::async_trait(?Send)]
+impl Perform for CreateReport {
+  type Response = CreateReportResponse;
+
+  async fn perform(
+    &self,
+    context: &Data<LemmyContext>,
+    websocket_id: Option<ConnectionId>,
+  ) -> Result<CreateReportResponse, LemmyError> {
+    let data: &CreateReport = &self;
+    let user = get_user_from_jwt(&data.auth, context.pool()).await?;
+
+    // check size of report and check for whitespace
+    let reason = data.reason.clone();
+    if reason.trim().is_empty() {
+      return Err(APIError::err("report_reason_required").into());
+    }
+    if reason.len() > MAX_REPORT_LEN {
+      return Err(APIError::err("report_too_long").into());
+    }
+
+    let report_type = ReportType::from_str(&data.report_type)?;
+    let user_id = user.id;
+    match report_type {
+      ReportType::Comment => { create_comment_report(context, data, user_id).await?; }
+      ReportType::Post => { create_post_report(context, data, user_id).await?; }
+    }
+
+    // to build on this, the user should get a success response, however
+    // mods should get a different response with more details
+    let res = CreateReportResponse { success: true };
+
+    context.chat_server().do_send(SendUserRoomMessage {
+      op: UserOperation::CreateReport,
+      response: res.clone(),
+      recipient_id: user.id,
+      websocket_id,
+    });
+
+    Ok(res)
+  }
+}
+
+#[async_trait::async_trait(?Send)]
+impl Perform for GetReportCount {
+  type Response = GetReportCountResponse;
+
+  async fn perform(
+    &self,
+    context: &Data<LemmyContext>,
+    websocket_id: Option<ConnectionId>,
+  ) -> Result<GetReportCountResponse, LemmyError> {
+    let data: &GetReportCount = &self;
+    let user = get_user_from_jwt(&data.auth, context.pool()).await?;
+    let community_id = data.community;
+
+    // Check for mod/admin privileges
+    is_mod_or_admin(context.pool(), user.id, community_id).await?;
+
+    let comment_reports = blocking(context.pool(), move |conn| {
+      CommentReportQueryBuilder::create(conn)
+        .community_id(community_id)
+        .resolved(false)
+        .count()
```

Comment:
It would be more efficient to create a special diesel thing that just does `select count(*) from comment_report where resolved is false`, because building the query might have additional sorts / query building work that isn't necessary.